File: day-6/main.py
import sys

# note I'm keeping this file, but the cpp file basically replaces it because it is faster
# and keep up w/ performance reqs
class LanternFish:
    def __init__(self, timer):
        self.timer = int(timer)
    
    def __repr__(self):
        return f'{self.timer}'

class LanternFishSim:
    def __init__(self, fish_timers: list):
        self.fishes = [LanternFish(timer) for timer in fish_timers]
        self.days = 0

    @staticmethod
    def init_from_file(file_name):
        with open(file_name) as f:
            # file input should only be one line
            return LanternFishSim(f.readline().split(','))

    def simulate_day(self):
        fishes_to_add = 0
        for fish in self.fishes:
            if fish.timer == 0:
                fish.timer = 6
                fishes_to_add += 1
            else:
                fish.timer -= 1
        for i in range(fishes_to_add):
            self.fishes.append(LanternFish(8))
        self.days += 1
        return self.fishes

    def simulate_days(self, days):
        for _ in range(days):
            self.simulate_day()
        return self.fishes

def part_one(file_name):
    sim = LanternFishSim.init_from_file(file_name)
    # print(f'Initial State: {sim.state(0)}')

    # for day in range(1,18):
    #     print(f'After {day} day{"s" if day>1 else ""}: {sim.state(day)}')
    print(f'Part One: Number of fish after 80 days: {len(sim.simulate_days(80))}')

def part_two(file_name):
    sim = LanternFishSim.init_from_file(file_name)
    print(f'Part Two: Number of fish after 256 days: {len(sim.simulate_days(256))}')
                

def main(args):
    if len(args) != 2:
        print('need file input provided as cli arg')
        return
    
    part_one(args[1])
    part_two(args[1])

if __name__ == '__main__':
    main(sys.argv)

File: day-6/main.cpp
# https://adventofcode.com/2021/day/6

#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <array>

class LanternFishSim {
public:
    static inline const LanternFishSim init_from_file(const std::string& file_name) {
        std::ifstream file(file_name);
        std::string data = "";
        std::vector<int> fish_timers;
        while(getline(file, data,','))
        {
           fish_timers.push_back(stoi(data)); 
        }
        return LanternFishSim(fish_timers);
    }

    LanternFishSim(const std::vector<int>& fish_timers) {
        for (const auto& fish_timer : fish_timers) {
            fish_timer_counts[fish_timer]+=1;
        }
        display_counts();
    }

    void display_counts() {
        for (size_t i { 0 }; i < 9; ++i) {
            std::cout << "Count for " << i << " timer: " << fish_timer_counts[i] << std::endl;
        }
    }

    void simulate_day() {
        std::cout << "Simulating day " << days << std::endl;
        size_t breeding_fish = fish_timer_counts[0];
        fish_timer_counts[0] = fish_timer_counts[1];
        fish_timer_counts[1] = fish_timer_counts[2];
        fish_timer_counts[2] = fish_timer_counts[3];
        fish_timer_counts[3] = fish_timer_counts[4];
        fish_timer_counts[4] = fish_timer_counts[5];
        fish_timer_counts[5] = fish_timer_counts[6];
        fish_timer_counts[6] = breeding_fish + fish_timer_counts[7];
        fish_timer_counts[7] = fish_timer_counts[8];
        fish_timer_counts[8] = breeding_fish;
        days += 1;
    }

    void simulate_days(const size_t days) {
        for (size_t i { 0 }; i < days; ++i) {
            simulate_day();
        }
    }

    size_t fish_count() { 
        size_t fish_count { 0 };
        for (const auto& count : fish_timer_counts) {
            fish_count += count;
        }
        return fish_count;
    }

private:
    std::array<size_t, 9> fish_timer_counts {0, 0, 0, 0, 0, 0, 0, 0, 0};
    size_t days { 0 };
};

void part_one(const std::string& file_name) {
    LanternFishSim sim { LanternFishSim::init_from_file(file_name) };
    sim.simulate_days(80);
    std::cout << "Part One: Number of fish after 80 days: " << sim.fish_count() << std::endl;
}

void part_two(const std::string& file_name) {
    LanternFishSim sim { LanternFishSim::init_from_file(file_name) };
    sim.simulate_days(256);
    std::cout << "Part One: Number of fish after 256 days: " << sim.fish_count() << std::endl;
}

int main(int argc, char** argv) {
    if (argc != 2) {
        std::cout << "Need to provide file input" << std::endl;
        return 0;
    }
    part_one(argv[1]);
    part_two(argv[1]);
}